src/sources/http.rs

```diff
@@ -0,0 +1,344 @@
+    fn build_event(
+        &self,
+        body: impl Buf,
+        header_map: HeaderMap,
+    ) -> Result<Vec<Event>, ErrorMessage> {
+        decode_body(body, self.encoding.clone())
+            .map(|events| add_headers(events, &self.headers, header_map))
+    }
+}
+
+#[typetag::serde(name = "http")]
+impl SourceConfig for SimpleHttpConfig {
+    fn build(
+        &self,
+        _: &str,
+        _: &GlobalOptions,
+        out: mpsc::Sender<Event>,
+    ) -> crate::Result<super::Source> {
+        let source = SimpleHttpSource {
+            encoding: self.encoding.clone(),
+            headers: self.headers.clone(),
+        };
+        source.run(self.address, "".to_string(), out)
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn source_type(&self) -> &'static str {
+        "http"
+    }
+}
+
+fn add_headers(
+    mut events: Vec<Event>,
+    headers_config: &Vec<String>,
+    headers: HeaderMap,
+) -> Vec<Event> {
+    for header_name in headers_config {
+        let mut value: &[u8] = &[];
+        if headers.contains_key(header_name) {
+            value = headers[header_name].as_bytes();
+        }
+        for event in events.iter_mut() {
+            event.as_mut_log().insert(header_name as &str, value);
+        }
+    }
+
+    events
+}
+
+fn decode_body(body: impl Buf, enc: Encoding) -> Result<Vec<Event>, ErrorMessage> {
+    let mut rdr = BufReader::new(body.reader())
+        .lines()
+        .filter_map(|res| {
+            res.map_err(|error| error!(message = "Error reading request body", ?error))
+                .ok()
+        })
+        .filter(|s| !s.is_empty());
+
+    match enc {
+        Encoding::Text => Ok(rdr.map(Event::from).collect()),
+        Encoding::Ndjson => Ok(rdr
+            .map(|j| json_parse_object(json_parse(j)?))
+            .collect::<Result<_, _>>()?),
+        Encoding::Json => {
+            let first_line = rdr.next();
+            let second_line = rdr.next();
+            if first_line.is_some() && second_line.is_none() {
+                //need exactly one line
+                json_parse_array_of_object(json_parse(first_line.unwrap())?)
+            } else {
+                Err(json_error())
+            }
+        }
+    }
+}
+
+fn json_parse(json_str: String) -> Result<JsonValue, ErrorMessage> {
+    serde_json::from_str(&json_str).map_err(|_| json_error())
+}
+
+fn json_parse_object(value: JsonValue) -> Result<Event, ErrorMessage> {
+    let mut event = Event::new_empty_log();
+    let log = event.as_mut_log();
+    log.insert(event::TIMESTAMP.clone(), Utc::now()); // Add timestamp
+    match value {
+        JsonValue::Object(map) => {
+            flatten(log, map);
+            Ok(event)
+        }
+        _ => Err(json_error()),
+    }
+}
+
+fn json_parse_array_of_object(value: JsonValue) -> Result<Vec<Event>, ErrorMessage> {
+    match value {
+        JsonValue::Array(v) => {
+            let mut out_vec = vec![];
```

Comment:
```rust
        JsonValue::Array(v) => Ok(v
            .into_iter()
            .map(json_parse_object)
            .collect::<Result<Vec<Event>, _>>()?),
```

There's a `FromIterator<Result<A, E>>` that makes this `collect()` call work. Kinda like `traverse` from Haskell. I didn't run any tests but I think this says the same thing.